File: sudoq-app/sudoqmodel/src/main/java/de/sudoq/model/solverGenerator/solution/NakedSetDerivation.java
package de.sudoq.model.solverGenerator.solution;

import java.util.ArrayList;
import java.util.List;
import java.util.Stack;

import de.sudoq.model.actionTree.Action;
import de.sudoq.model.actionTree.NoteActionFactory;
import de.sudoq.model.solvingAssistant.HintTypes;
import de.sudoq.model.sudoku.CandidateSet;
import de.sudoq.model.sudoku.Constraint;
import de.sudoq.model.sudoku.Sudoku;

/**
 * Created by timo on 04.10.16.
 */
public class NakedSetDerivation extends SolveDerivation {


    private Constraint constraint;
    private List<DerivationCell> subsetMembers;
    private List<DerivationCell> externalCells;
    private CandidateSet subsetCandidates;

    public NakedSetDerivation(HintTypes technique) {
        super(technique);
        subsetMembers  = new Stack<>();
        externalCells = new Stack<>();
        hasActionListCapability = true;
    }

    public void setSubsetCandidates(CandidateSet bs){
        subsetCandidates = (CandidateSet) bs.clone();
    }

    public void setConstraint(Constraint c){ constraint = c; }

    public void addExternalCell(DerivationCell f){
        externalCells.add(f);
    }

    public void addSubsetCell(DerivationCell f){
        subsetMembers.add(f);
    }


    public CandidateSet getSubsetCandidates(){ return subsetCandidates; }

    public Constraint getConstraint(){ return constraint; }

    public List<DerivationCell> getSubsetMembers(){ return subsetMembers; }

    public List<DerivationCell> getExternalCellsMembers(){ return externalCells; }

    /* creates a list of actions in case the user want the app to execute the hints */
    @Override
    public List<Action> getActionList(Sudoku sudoku){
        List<Action> actionlist = new ArrayList<>();
        NoteActionFactory af = new NoteActionFactory();

        for(DerivationCell df : externalCells)
            for(int note: CandidateSet.fromBitSet(df.getRelevantCandidates()).getSetBits())
                actionlist.add(af.createAction(note, sudoku.getCell(df.getPosition())));

        return actionlist;
    }


}

File: sudoq-app/sudoqmodel/src/main/kotlin/de/sudoq/model/solverGenerator/solution/NakedSetDerivation.kt
package de.sudoq.model.solverGenerator.solution

import de.sudoq.model.actionTree.Action
import de.sudoq.model.actionTree.NoteActionFactory
import de.sudoq.model.solvingAssistant.HintTypes
import de.sudoq.model.sudoku.CandidateSet
import de.sudoq.model.sudoku.CandidateSet.Companion.fromBitSet
import de.sudoq.model.sudoku.Constraint
import de.sudoq.model.sudoku.Sudoku
import java.util.Stack

/**
 * Created by timo on 04.10.16.
 */
class NakedSetDerivation(technique: HintTypes) : SolveDerivation(technique) {
    var constraint: Constraint? = null
    private val subsetMembers: MutableList<DerivationCell>
    private val externalCells: MutableList<DerivationCell>
    var subsetCandidates: CandidateSet? = null
        private set

    fun setSubsetCandidates(bs: CandidateSet) {
        subsetCandidates = bs.clone() as CandidateSet
    }

    fun addExternalCell(f: DerivationCell) {
        externalCells.add(f)
    }

    fun addSubsetCell(f: DerivationCell) {
        subsetMembers.add(f)
    }

    fun getSubsetMembers(): List<DerivationCell> {
        return subsetMembers
    }

    val externalCellsMembers: List<DerivationCell>
        get() = externalCells

    /* creates a list of actions in case the user want the app to execute the hints */
    override fun getActionList(sudoku: Sudoku): List<Action> {
        val actionlist: MutableList<Action> = ArrayList()
        val af = NoteActionFactory()
        for (df in externalCells) for (note in fromBitSet(df.relevantCandidates).setBits) actionlist.add(
            af.createAction(note, sudoku.getCell(df.position))
        )
        return actionlist
    }

    init {
        subsetMembers = Stack()
        externalCells = Stack()
        hasActionListCapability = true
    }
}